File: Sources/AgentManagerCore/Workflow/WorkflowStore.swift
import Foundation
import Yams

public struct WorkflowStore {
    public static let shared = WorkflowStore()

    public let baseDirectory: URL
    public let workflowsDirectory: URL
    public let runsDirectory: URL

    public init(baseDirectory: URL? = nil) {
        let home = FileManager.default.homeDirectoryForCurrentUser
        self.baseDirectory = baseDirectory ?? home.appendingPathComponent(".agent-manager")
        self.workflowsDirectory = self.baseDirectory.appendingPathComponent("workflows")
        self.runsDirectory = self.baseDirectory.appendingPathComponent("runs")
    }

    /// Ensure all required directories exist
    public func ensureDirectoriesExist() throws {
        let fm = FileManager.default
        try fm.createDirectory(at: workflowsDirectory, withIntermediateDirectories: true)
        try fm.createDirectory(at: runsDirectory, withIntermediateDirectories: true)
    }

    // MARK: - Workflow Path Helpers

    /// Get path to workflow YAML file
    public func workflowPath(name: String) -> URL {
        workflowsDirectory.appendingPathComponent("\(name).yaml")
    }

    /// Get path to workflow's run directory
    public func runDirectory(workflowName: String) -> URL {
        runsDirectory.appendingPathComponent(workflowName)
    }

    /// Get path to a specific run file
    public func runPath(workflowName: String, runId: String) -> URL {
        runDirectory(workflowName: workflowName)
            .appendingPathComponent("\(runId).json")
    }

    /// Get path to run's log directory
    public func runLogDirectory(workflowName: String, runId: String) -> URL {
        runDirectory(workflowName: workflowName)
            .appendingPathComponent(runId)
    }

    // MARK: - Workflow CRUD Operations

    /// List all workflow names
    public func listWorkflowNames() throws -> [String] {
        let fm = FileManager.default

        guard fm.fileExists(atPath: workflowsDirectory.path) else {
            return []
        }

        let files = try fm.contentsOfDirectory(at: workflowsDirectory, includingPropertiesForKeys: nil)
        return files
            .filter { $0.pathExtension == "yaml" || $0.pathExtension == "yml" }
            .map { $0.deletingPathExtension().lastPathComponent }
            .sorted()
    }

    /// Load all workflows
    public func listWorkflows() throws -> [Workflow] {
        let names = try listWorkflowNames()
        return names.compactMap { name in
            try? load(name: name)
        }
    }

    /// Load a specific workflow by name
    public func load(name: String) throws -> Workflow {
        let path = workflowPath(name: name)

        guard FileManager.default.fileExists(atPath: path.path) else {
            throw WorkflowStoreError.workflowNotFound(name)
        }

        return try Workflow.load(fromFile: path.path)
    }

    /// Check if a workflow exists
    public func exists(name: String) -> Bool {
        FileManager.default.fileExists(atPath: workflowPath(name: name).path)
    }

    /// Save a workflow
    public func save(_ workflow: Workflow) throws {
        try ensureDirectoriesExist()

        // Validate before saving
        try workflow.validate()

        let path = workflowPath(name: workflow.name)
        try workflow.save(toFile: path.path)
    }

    /// Delete a workflow
    public func delete(name: String) throws {
        let path = workflowPath(name: name)

        guard FileManager.default.fileExists(atPath: path.path) else {
            throw WorkflowStoreError.workflowNotFound(name)
        }

        try FileManager.default.removeItem(at: path)
    }

    /// Get workflow info for display
    public func getWorkflowInfo(name: String) throws -> WorkflowInfo {
        let workflow = try load(name: name)
        let isEnabled = WorkflowLaunchAgentManager.shared.isInstalled(workflowName: name)
        let stats = try? loadWorkflowStats(name: name)
        let lastRun = try? loadLastRun(workflowName: name)

        return WorkflowInfo(
            name: workflow.name,
            description: workflow.description,
            jobCount: workflow.jobs.count,
            hasSchedule: workflow.on.hasSchedule,
            isEnabled: isEnabled,
            lastRunStatus: lastRun?.status,
            lastRunDate: lastRun?.endTime ?? lastRun?.startTime,
            stats: stats
        )
    }

    /// Get info for all workflows
    public func listWorkflowInfo() throws -> [WorkflowInfo] {
        let names = try listWorkflowNames()
        return names.compactMap { name in
            try? getWorkflowInfo(name: name)
        }
    }

    // MARK: - Run Operations

    /// Save a workflow run
    public func saveRun(_ run: WorkflowRun) throws {
        try ensureDirectoriesExist()

        let runDir = runDirectory(workflowName: run.workflowName)
        try FileManager.default.createDirectory(at: runDir, withIntermediateDirectories: true)

        let runPath = self.runPath(workflowName: run.workflowName, runId: run.id)

        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        let data = try encoder.encode(run)
        try data.write(to: runPath)

        // Update workflow stats
        try updateWorkflowStats(name: run.workflowName, run: run)
    }

    /// Load a specific run
    public func loadRun(workflowName: String, runId: String) throws -> WorkflowRun {
        let path = runPath(workflowName: workflowName, runId: runId)

        guard FileManager.default.fileExists(atPath: path.path) else {
            throw WorkflowStoreError.runNotFound(runId)
        }

        let data = try Data(contentsOf: path)
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return try decoder.decode(WorkflowRun.self, from: data)
    }

    /// List runs for a workflow (most recent first)
    public func listRuns(workflowName: String, limit: Int = 10) throws -> [WorkflowRun] {
        let runDir = runDirectory(workflowName: workflowName)

        guard FileManager.default.fileExists(atPath: runDir.path) else {
            return []
        }

        let files = try FileManager.default.contentsOfDirectory(at: runDir, includingPropertiesForKeys: [.contentModificationDateKey])
            .filter { $0.pathExtension == "json" }
            .sorted { url1, url2 in
                let date1 = (try? url1.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate) ?? Date.distantPast
                let date2 = (try? url2.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate) ?? Date.distantPast
                return date1 > date2
            }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601

        return files.prefix(limit).compactMap { url -> WorkflowRun? in
            guard let data = try? Data(contentsOf: url),
                  let run = try? decoder.decode(WorkflowRun.self, from: data) else {
                return nil
            }
            return run
        }
    }

    /// Get the most recent run for a workflow
    public func loadLastRun(workflowName: String) throws -> WorkflowRun? {
        let runs = try listRuns(workflowName: workflowName, limit: 1)
        return runs.first
    }

    /// Delete old runs (keep most recent N)
    public func pruneRuns(workflowName: String, keepCount: Int = 50) throws {
        let runDir = runDirectory(workflowName: workflowName)

        guard FileManager.default.fileExists(atPath: runDir.path) else {
            return
        }

        let files = try FileManager.default.contentsOfDirectory(at: runDir, includingPropertiesForKeys: [.contentModificationDateKey])
            .filter { $0.pathExtension == "json" }
            .sorted { url1, url2 in
                let date1 = (try? url1.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate) ?? Date.distantPast
                let date2 = (try? url2.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate) ?? Date.distantPast
                return date1 > date2
            }

        // Delete runs beyond the keep count
        for url in files.dropFirst(keepCount) {
            try? FileManager.default.removeItem(at: url)

            // Also delete the log directory if it exists
            let logDir = url.deletingPathExtension()
            if FileManager.default.fileExists(atPath: logDir.path) {
                try? FileManager.default.removeItem(at: logDir)
            }
        }
    }

    // MARK: - Stats Operations

    private var statsFile: URL {
        baseDirectory.appendingPathComponent("workflow-stats.json")
    }

    /// Load all workflow stats
    public func loadAllWorkflowStats() throws -> [String: WorkflowStats] {
        guard FileManager.default.fileExists(atPath: statsFile.path) else {
            return [:]
        }

        let data = try Data(contentsOf: statsFile)
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return try decoder.decode([String: WorkflowStats].self, from: data)
    }

    /// Load stats for a specific workflow
    public func loadWorkflowStats(name: String) throws -> WorkflowStats {
        let allStats = try loadAllWorkflowStats()
        return allStats[name] ?? WorkflowStats()
    }

    /// Save workflow stats
    private func saveWorkflowStats(_ stats: [String: WorkflowStats]) throws {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        encoder.outputFormatting = .prettyPrinted
        let data = try encoder.encode(stats)
        try data.write(to: statsFile)
    }

    /// Update stats for a workflow after a run
    private func updateWorkflowStats(name: String, run: WorkflowRun) throws {
        var allStats = (try? loadAllWorkflowStats()) ?? [:]
        var stats = allStats[name] ?? WorkflowStats()
        stats.recordRun(run)
        allStats[name] = stats
        try saveWorkflowStats(allStats)
    }

    /// Delete stats for a workflow
    public func deleteWorkflowStats(name: String) throws {
        var allStats = (try? loadAllWorkflowStats()) ?? [:]
        allStats.removeValue(forKey: name)
        try saveWorkflowStats(allStats)
    }
}

// MARK: - WorkflowInfo

public struct WorkflowInfo {
    public let name: String
    public let description: String?
    public let jobCount: Int
    public let hasSchedule: Bool
    public let isEnabled: Bool
    public let lastRunStatus: WorkflowStatus?
    public let lastRunDate: Date?
    public let stats: WorkflowStats?

    public init(
        name: String,
        description: String?,
        jobCount: Int,
        hasSchedule: Bool,
        isEnabled: Bool,
        lastRunStatus: WorkflowStatus?,
        lastRunDate: Date?,
        stats: WorkflowStats?
    ) {
        self.name = name
        self.description = description
        self.jobCount = jobCount
        self.hasSchedule = hasSchedule
        self.isEnabled = isEnabled
        self.lastRunStatus = lastRunStatus
        self.lastRunDate = lastRunDate
        self.stats = stats
    }

    public var statusIndicator: String {
        if isEnabled {
            return "●"
        } else {
            return "○"
        }
    }

    public var lastRunStatusIcon: String? {
        guard let status = lastRunStatus else { return nil }
        switch status {
        case .completed: return "✓"
        case .failed: return "✗"
        case .running: return "◐"
        case .cancelled: return "◌"
        case .pending: return "○"
        }
    }
}

// MARK: - Errors

public enum WorkflowStoreError: Error, LocalizedError {
    case workflowNotFound(String)
    case workflowAlreadyExists(String)
    case invalidWorkflowName(String)
    case runNotFound(String)
    case noScheduleTrigger(String)
    case agentNotFound(workflowName: String, agentName: String)

    public var errorDescription: String? {
        switch self {
        case .workflowNotFound(let name):
            return "Workflow '\(name)' not found"
        case .workflowAlreadyExists(let name):
            return "Workflow '\(name)' already exists"
        case .invalidWorkflowName(let name):
            return "Invalid workflow name: '\(name)'"
        case .runNotFound(let id):
            return "Workflow run '\(id)' not found"
        case .noScheduleTrigger(let name):
            return "Workflow '\(name)' has no schedule trigger configured"
        case .agentNotFound(let workflowName, let agentName):
            return "Workflow '\(workflowName)' references unknown agent '\(agentName)'"
        }
    }
}

File: agent_manager/store/workflow_store.py
"""Workflow storage and management."""

from __future__ import annotations

import json
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path

from agent_manager.models.result import WorkflowRun, WorkflowStats, WorkflowStatus
from agent_manager.models.workflow import Workflow


class WorkflowStoreError(Exception):
    """Base exception for workflow store errors."""

    pass


class WorkflowNotFoundError(WorkflowStoreError):
    """Workflow not found."""

    def __init__(self, name: str):
        self.name = name
        super().__init__(f"Workflow '{name}' not found")


class WorkflowAlreadyExistsError(WorkflowStoreError):
    """Workflow already exists."""

    def __init__(self, name: str):
        self.name = name
        super().__init__(f"Workflow '{name}' already exists")


class RunNotFoundError(WorkflowStoreError):
    """Workflow run not found."""

    def __init__(self, run_id: str):
        self.run_id = run_id
        super().__init__(f"Workflow run '{run_id}' not found")


@dataclass
class WorkflowInfo:
    """Summary information about a workflow."""

    name: str
    description: str | None
    job_count: int
    has_schedule: bool
    is_enabled: bool
    last_run_status: WorkflowStatus | None = None
    last_run_date: datetime | None = None
    stats: WorkflowStats | None = None

    @property
    def status_indicator(self) -> str:
        """Status indicator character."""
        return "●" if self.is_enabled else "○"

    @property
    def last_run_status_icon(self) -> str | None:
        """Icon for last run status."""
        if self.last_run_status is None:
            return None
        icons = {
            WorkflowStatus.COMPLETED: "✓",
            WorkflowStatus.FAILED: "✗",
            WorkflowStatus.RUNNING: "◐",
            WorkflowStatus.CANCELLED: "◌",
            WorkflowStatus.PENDING: "○",
        }
        return icons.get(self.last_run_status)


class WorkflowStore:
    """Manages workflow persistence and run history."""

    def __init__(self, base_directory: Path | None = None):
        if base_directory is None:
            base_directory = Path.home() / ".agent-manager"

        self.base_directory = base_directory
        self.workflows_directory = base_directory / "workflows"
        self.runs_directory = base_directory / "runs"
        self.stats_file = base_directory / "workflow-stats.json"

    def ensure_directories_exist(self) -> None:
        """Ensure all required directories exist."""
        self.workflows_directory.mkdir(parents=True, exist_ok=True)
        self.runs_directory.mkdir(parents=True, exist_ok=True)

    # MARK: - Workflow Path Helpers

    def workflow_path(self, name: str) -> Path:
        """Get path to workflow YAML file."""
        return self.workflows_directory / f"{name}.yaml"

    def run_directory(self, workflow_name: str) -> Path:
        """Get path to workflow's run directory."""
        return self.runs_directory / workflow_name

    def run_path(self, workflow_name: str, run_id: str) -> Path:
        """Get path to a specific run file."""
        return self.run_directory(workflow_name) / f"{run_id}.json"

    def run_log_directory(self, workflow_name: str, run_id: str) -> Path:
        """Get path to run's log directory."""
        return self.run_directory(workflow_name) / run_id

    # MARK: - Workflow CRUD Operations

    def list_workflow_names(self) -> list[str]:
        """List all workflow names."""
        if not self.workflows_directory.exists():
            return []

        return sorted([
            p.stem
            for p in self.workflows_directory.iterdir()
            if p.suffix in (".yaml", ".yml")
        ])

    def list_workflows(self) -> list[Workflow]:
        """Load all workflows."""
        workflows = []
        for name in self.list_workflow_names():
            try:
                workflows.append(self.load(name))
            except Exception:
                pass  # Skip invalid workflows
        return workflows

    def load(self, name: str) -> Workflow:
        """Load a specific workflow by name."""
        path = self.workflow_path(name)

        if not path.exists():
            raise WorkflowNotFoundError(name)

        return Workflow.load(path)

    def exists(self, name: str) -> bool:
        """Check if a workflow exists."""
        return self.workflow_path(name).exists()

    def save(self, workflow: Workflow) -> None:
        """Save a workflow."""
        self.ensure_directories_exist()
        path = self.workflow_path(workflow.name)
        workflow.save(path)

    def delete(self, name: str) -> None:
        """Delete a workflow."""
        path = self.workflow_path(name)

        if not path.exists():
            raise WorkflowNotFoundError(name)

        path.unlink()

    def get_workflow_info(self, name: str) -> WorkflowInfo:
        """Get workflow info for display."""
        workflow = self.load(name)
        # TODO: Check LaunchAgent status
        is_enabled = False  # Placeholder

        stats = self.load_workflow_stats(name)
        last_run = self.load_last_run(name)

        return WorkflowInfo(
            name=workflow.name,
            description=workflow.description,
            job_count=len(workflow.jobs),
            has_schedule=workflow.on.has_schedule,
            is_enabled=is_enabled,
            last_run_status=last_run.status if last_run else None,
            last_run_date=last_run.end_time or last_run.start_time if last_run else None,
            stats=stats if stats.total_runs > 0 else None,
        )

    def list_workflow_info(self) -> list[WorkflowInfo]:
        """Get info for all workflows."""
        infos = []
        for name in self.list_workflow_names():
            try:
                infos.append(self.get_workflow_info(name))
            except Exception:
                pass
        return infos

    # MARK: - Run Operations

    def save_run(self, run: WorkflowRun) -> None:
        """Save a workflow run."""
        self.ensure_directories_exist()

        run_dir = self.run_directory(run.workflow_name)
        run_dir.mkdir(parents=True, exist_ok=True)

        run_path = self.run_path(run.workflow_name, run.id)
        run_path.write_text(run.model_dump_json(by_alias=True, indent=2))

        # Update workflow stats
        self._update_workflow_stats(run.workflow_name, run)

    def load_run(self, workflow_name: str, run_id: str) -> WorkflowRun:
        """Load a specific run."""
        path = self.run_path(workflow_name, run_id)

        if not path.exists():
            raise RunNotFoundError(run_id)

        data = json.loads(path.read_text())
        return WorkflowRun.model_validate(data)

    def list_runs(self, workflow_name: str, limit: int = 10) -> list[WorkflowRun]:
        """List runs for a workflow (most recent first)."""
        run_dir = self.run_directory(workflow_name)

        if not run_dir.exists():
            return []

        # Get JSON files sorted by modification time (newest first)
        run_files = sorted(
            [p for p in run_dir.iterdir() if p.suffix == ".json"],
            key=lambda p: p.stat().st_mtime,
            reverse=True,
        )

        runs = []
        for path in run_files[:limit]:
            try:
                data = json.loads(path.read_text())
                runs.append(WorkflowRun.model_validate(data))
            except Exception:
                pass  # Skip invalid runs

        return runs

    def load_last_run(self, workflow_name: str) -> WorkflowRun | None:
        """Get the most recent run for a workflow."""
        runs = self.list_runs(workflow_name, limit=1)
        return runs[0] if runs else None

    def prune_runs(self, workflow_name: str, keep_count: int = 50) -> None:
        """Delete old runs (keep most recent N)."""
        run_dir = self.run_directory(workflow_name)

        if not run_dir.exists():
            return

        # Get JSON files sorted by modification time (newest first)
        run_files = sorted(
            [p for p in run_dir.iterdir() if p.suffix == ".json"],
            key=lambda p: p.stat().st_mtime,
            reverse=True,
        )

        # Delete runs beyond the keep count
        for path in run_files[keep_count:]:
            path.unlink(missing_ok=True)

            # Also delete the log directory if it exists
            log_dir = path.with_suffix("")
            if log_dir.is_dir():
                import shutil

                shutil.rmtree(log_dir, ignore_errors=True)

    # MARK: - Stats Operations

    def load_all_workflow_stats(self) -> dict[str, WorkflowStats]:
        """Load all workflow stats."""
        if not self.stats_file.exists():
            return {}

        data = json.loads(self.stats_file.read_text())
        return {
            name: WorkflowStats.model_validate(stats) for name, stats in data.items()
        }

    def load_workflow_stats(self, name: str) -> WorkflowStats:
        """Load stats for a specific workflow."""
        all_stats = self.load_all_workflow_stats()
        return all_stats.get(name, WorkflowStats())

    def _save_workflow_stats(self, stats: dict[str, WorkflowStats]) -> None:
        """Save workflow stats."""
        data = {name: s.model_dump(by_alias=True) for name, s in stats.items()}
        self.stats_file.parent.mkdir(parents=True, exist_ok=True)
        self.stats_file.write_text(json.dumps(data, indent=2, default=str))

    def _update_workflow_stats(self, name: str, run: WorkflowRun) -> None:
        """Update stats for a workflow after a run."""
        all_stats = self.load_all_workflow_stats()
        stats = all_stats.get(name, WorkflowStats())
        stats.record_run(run)
        all_stats[name] = stats
        self._save_workflow_stats(all_stats)

    def delete_workflow_stats(self, name: str) -> None:
        """Delete stats for a workflow."""
        all_stats = self.load_all_workflow_stats()
        all_stats.pop(name, None)
        self._save_workflow_stats(all_stats)


# Singleton instance
_store: WorkflowStore | None = None


def get_workflow_store() -> WorkflowStore:
    """Get the singleton workflow store instance."""
    global _store
    if _store is None:
        _store = WorkflowStore()
    return _store
